Add task counter

nightism opened the pull request:
![image](https://cloud.githubusercontent.com/assets/19910007/20038232/ef893870-a46a-11e6-864f-9ffadf4ed9c5.png)

Task Counters are added:
1. total task counter: all tasks stored in database
2. listed task counter: count the tasks listed in ui
3. unfinished task counter: count the unfinished tasks stored in database
4. overdue task counter: count the overdue deadline tasks stored in database
5. today's remaining task counter: count the remaining deadlines and events for today

Commit:
FootBar: Add total Task counter

Commit:
TaskCounter: add task counter to evnet task

Commit:
TaskCounter: add task counter to deadline task

Commit:
TaskCounter: add task counter to floating task

Commit:
FootBar: TotalTaskCounter revised

* now total task counter counts the total number of the task stored in TaskTracker rather than the tasks listed in UI

Commit:
TaskCounter: add finished floating tasks counter

Commit:
TaskCounter: add finished deadline counter

Commit:
TaskCounter: add finished event counter

Commit:
TaskCounter: add overdue deadline counter

Commit:
Ui: polish ui

Commit:
TaskCounter: add today remaining counter

pyokagan commented on src/main/java/seedu/address/ui/DeadlineTaskListPane.java:
```diff
@@ -17,10 +23,27 @@
         deadlineTaskListView.setCellFactory(listView -> new DeadlineTaskListCell());
+        listedDeadlineCounter.textProperty().bind(Bindings.size(model.getDeadlineTaskList())
+                                                  .asString("Number of Deadlines listed: %d"));
+        todayUnfinishedDeadlineCounter.textProperty().bind(Bindings.size(model.getDeadlineTaskList(new TaskWillHappenTodayPredicate(LocalDateTime.now())))
```
One major problem though: the `TaskWillHappenTodayPredicate` will always compare with the time at which it was created. In order words: the counter will not change with time.

nightism replied:
Emmm... yes... just realise. Do you have any suggestions?

nightism replied:
Since we do not want to use event bus, so how do we actually tell the time when task book changes?

pyokagan replied:
`ObservableList.addListener()`?

Commit:
Filter: update TaskWillHappenTodayPredicate

Commit:
EventTaskCounter: reset the filters used in TaskCounter when tasklist is updated

Now when the event tasks listed in the ui change, the filters used when generating task counter will be updated with the last modified time

Commit:
no message

Commit:
DeadlineTaskCounter: reset the filters used in TaskCounter when tasklist is updated

Now when the deadline tasks listed in the ui change, the filters used when generating task counter will be updated with the last modified time

Commit:
small changes in FloatingTaskListPane

This commit is to make FloatingTaskListPane similar to Deadline/EventTaskListPane

nightism commented:
@pyokagan Please help me review this pr again, thanks!

nightism commented:
@pyokagan Can this pr be merged?

pyokagan commented on src/main/java/seedu/address/ui/DeadlineTaskListPane.java:
```diff
@@ -1,26 +1,71 @@
+
+        // Initialize task lists
+        this.listedDeadlineTasks = model.getDeadlineTaskList();
+        this.todayDeadlineTasks = model.getDeadlineTaskList(new TaskWillHappenTodayPredicate(LocalDateTime.now()));
```
@nightism Should these counters be limited to what is actually shown in the UI or not? As of now, your `listedDeadlineTasks` is showing the filtered list, while `todayDeadlineTasks`, `overdueDeadlineTasks` is showing the unfiltered list!!!!

pyokagan replied:
e.g. 0 deadlines listed, but have overdue tasks and unfinished tasks????
![listed](https://cloud.githubusercontent.com/assets/109479/20046235/a1f6b90c-a4e2-11e6-93f4-ac99dccc1691.png)

nightism replied:
> As of now, your `listedDeadlineTasks` is showing the filtered list, while `todayDeadlineTasks`, `overdueDeadlineTasks` is showing the unfiltered list!!!!

Yes you are correct. But, I think that's why I indicate `Number of Task listed` for `listedDeadlineTasks` in the UI.

nightism replied:
If all perform on filtered list, then `todayDeadlineTasks`, `overdueDeadlineTasks` may not be very useful. Or we just delete `listedDeadlineTasks`.

nightism replied:
ok sure, will follow your advice

nightism replied:
will finish soon, please keep online... thanks very much

pyokagan replied:
Anyway, so I cleaned up your code (removed all the `Bindings` repetition and the not-required changes to the `Model` interface), and then made the counters only count the filtered list:

https://github.com/CS2103AUG2016-T11-C4/main/pull/185/commits/9267857320217880563145491260f30ab1cd2f73

What do you think?

pyokagan replied:
But yeah, (and I'll ask @potatowagon for a second opinion), any counters just above the panes of the filtered list should only count the filtered list items, or else I think users will be confused (or Damith will _think_ the users will be confused).

nightism replied:
@pyokagan ok, saw your revision already, it's ok

pyokagan replied:
@nightism ?? So do I merge that or not?

pyokagan replied:
OK I think I will just merge that one and credit you.

pyokagan replied:
Reason being: I will much rather we do not touch the `Model` interface any more. The resulting master..PR diff is also smaller, and we have zero code coverage drops with that.

But please got through it and check for bugs :)

pyokagan commented on src/main/java/seedu/address/model/ReadOnlyModel.java:
```diff
@@ -63,6 +66,9 @@
 
+    /** Return a filtered Deadline task list filter by given task predicate. */
+    ObservableList<DeadlineTask> getDeadlineTaskList(TaskPredicate taskFilter);
+
```
Also, this comment is wrong. This returns the underlying _unfiltered_ task list.

nightism replied:
ok, will revise the comment

Commit:
TaskCounter: now task counters only keep track on the listed tasks

nightism commented:
@pyokagan revision done already, now task counter only counts the task listed in the ui.
Reconsidering about it, I also think this way is less confusing.

pyokagan commented:
Merged as 3e6828bb6ab745503bd3185af801594ef58c0f6a
